bluemira/structural/transformation.py

```diff
@@ -0,0 +1,232 @@
+    u = np.array([dx, dy, dz])
+    x_local = u / np.linalg.norm(u)
+
+    x_global = np.array([1.0, 0, 0])
+    y_global = np.array([0, 1.0, 0])
+    z_global = np.array([0, 0, 1.0])
+    globa = np.array([x_global, y_global, z_global])
+
+    if x_local[0] == 1:
+        # corresponds to the global coordinate system
+        dcm = np.eye(3)
+        if debug:
+            return dcm, globa
+        else:
+            return dcm
+    if x_local[0] == -1:
+        # corresponds to the mirrored coordinate system
+        dcm = np.eye(3)
+        dcm[0, 0] = -1
+        local = globa.copy()
+        local[0][0] = -1
+        if debug:
+            return dcm, local
+        else:
+            return dcm
+    if abs(x_local[1]) == 1:
+        # corresponds to a local y-vector sitting on the global x-vector
+        # (rotation about z-axis)
+        cos_theta = 0
+        sin_theta = np.sign(x_local[1])
+        dcm = np.array(
+            [[cos_theta, -sin_theta, 0], [sin_theta, cos_theta, 0], [0, 0, 1]]
+        )
+        if debug:
+            local = np.array([[0, sin_theta, 0], [-sin_theta, 0, 0], [0, 0, 1]])
+            return dcm, local
+        else:
+            return dcm
+    if x_local[2] == 1:
+        # corresponds to a local z-vector sitting on the global x-vector
+        # rotation about the y-axis
+        cos_theta = 0
+        sin_theta = -1
+        dcm = np.array(
+            [[cos_theta, 0, sin_theta], [0, 1, 0], [-sin_theta, 0, cos_theta]]
+        )
+        if debug:
+            local = np.array([[0, 0, 1], [0, 1, 0], [-1, 0, 0]])
+            return dcm, local
+        else:
+            return dcm
+
+    y_local = np.array([0, 1.0, 0])
+    y_local -= np.dot(x_local, y_local) * x_local
+    y_local /= np.linalg.norm(y_local)
+    z_local = np.cross(x_local, y_local)
+
+    local = np.array([x_local, y_local, z_local])
+
+    if not isinstance(dcm, int):
+        if debug:
+            return dcm, local
+        else:
+            return dcm
+    c11 = np.dot(x_global, x_local)
+    c12 = np.dot(x_global, y_local)
+    c13 = np.dot(x_global, z_local)
+    c21 = np.dot(y_global, x_local)
+    c22 = np.dot(y_global, y_local)
+    c23 = np.dot(y_global, z_local)
+    c31 = np.dot(z_global, x_local)
+    c32 = np.dot(z_global, y_local)
+    c33 = np.dot(z_global, z_local)
+    dcm = np.array([[c11, c12, c13], [c21, c22, c23], [c31, c32, c33]])
+    if not debug:
+        return dcm
+    else:
+        return dcm, local
+
+
+def lambda_matrix(dx, dy, dz):
+    """
+    3-D transformation matrix, generalised to a 2-node beam.
+    Effectively only handles the tiling of the direction cosine matrix
+    transform.
+
+    Returns
+    -------
+    llambda: np.array((12, 12))
+        The transformation matrix
+    """
+    dcm = _direction_cosine_matrix(dx, dy, dz)
+    llambda = np.zeros((12, 12))
+    llambda[0:3, 0:3] = dcm
+    llambda[3:6, 3:6] = dcm
+    llambda[6:9, 6:9] = dcm
+    llambda[9:12, 9:12] = dcm
```

Comment:
all in one line possible I think with some fancy indexing or use of np.repeat?

Reply:
```python
for i in range(4):
    l2[i*3:(i+1)*3, i*3:(i+1)*3] = dcm
```
:S

Reply:
thats worse :smile:

Reply:
For future reference, `scipy.linalg` has [`block_diag`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.block_diag.html):

```python
llambda = block_diag(*[dcm]*4)
```

Reply:
Thats the sort of thing I was trying to find!